### PR TITLE
(PDB-3611) Deduplicate fact values (again)

### PR DESCRIPTION
**Pull request by mullr**
Previously, we had changed fact storage to put the value in the facts table, in
order to simplify storage and sidestep the problematic garbage collection issues
when fact values had their own table. But this proved to make queries too slow,
especially when dealing with queries which have multiple conditions on facts.

### CODE REVIEW

**Comment by mullr** [comment]
TODO:
- [ ] Shortcut the initial redupe migration for users that haven't done it yet
- [ ] Add a migration diff test

File: src/puppetlabs/puppetdb/scf/migrate.clj [hunk 1]
```diff
@@ -1281,13 +1283,100 @@
...
+   ;; values for this yet
+   "CREATE TABLE fact_values (
+      id bigint DEFAULT nextval('fact_values_id_seq'::regclass) NOT NULL,
+      value_hash bytea,
+      value_type_id bigint NOT NULL,
+      value_integer bigint,
+      value_float double precision,
+      value_string text,
+      value_boolean boolean,
+      value jsonb not null
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: do we not allow null values? I thought that used to be ok

File: src/puppetlabs/puppetdb/scf/migrate.clj [hunk 2]
```diff
@@ -1281,13 +1283,100 @@
...
+   "INSERT INTO facts_transform (factset_id, fact_path_id, fact_value_id)
+       SELECT f.factset_id, f.fact_path_id, fv.id
+         FROM facts f
+        INNER JOIN fact_values fv
+                ON fv.value_type_id = f.value_type_id
+               AND fv.value_integer IS NOT DISTINCT FROM f.value_integer
+               AND fv.value_float   IS NOT DISTINCT FROM f.value_float
+               AND fv.value_string  IS NOT DISTINCT FROM f.value_string
+               AND fv.value_boolean IS NOT DISTINCT FROM f.value_boolean
+               AND fv.value         = f.value")
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: if null values are allowed I think this needs to be "is not distinct from" as well

File: src/puppetlabs/puppetdb/scf/migrate.clj [hunk 3]
```diff
@@ -1281,13 +1283,100 @@
...
+
+  ;; populate fact_values.value_hash
+  (jdbc/call-with-query-rows
+   ["select id, value::text from fact_values"]
+   (fn [rows]
+     (doseq [batch (partition-all 5000 rows)]
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: we need to test this to make sure it won't oom with big facts

File: src/puppetlabs/puppetdb/scf/migrate.clj (hunk 2 above)
Comment (mullr) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: I had 'is not distinct from', but that made the query take... a really long time. (I ran it overnight with the orch scale dataset, and it didn't finish). That's why I added the not-null constraint above. I believe it's valid, as we have jsonb versions of all the values.

File: src/puppetlabs/puppetdb/scf/migrate.clj (hunk 3 above)
Comment (mullr) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: ... yes. What's the biggest fact you've heard of?

File: src/puppetlabs/puppetdb/scf/migrate.clj (hunk 2 above)
Comment (mullr) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: But the current failing unit tests are telling my this isn't a true assumption.

Comment (mullr) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: Even making this explicit, we end up with 

```
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Insert on facts_transform  (cost=332902.86..207785761576.67 rows=1 width=24)
   ->  Hash Join  (cost=332902.86..207785761576.67 rows=1 width=24)
         Hash Cond: (f.value_type_id = fv.value_type_id)
         Join Filter: (((fv.value = f.value) OR ((fv.value IS NULL) AND (f.value IS NULL))) AND (NOT (fv.value_integer IS DISTINCT FROM f.value_integer)) AND (NOT (fv.value_float IS
DISTINCT FROM f.value_float)) AND (NOT (fv.value_string IS DISTINCT FROM f.value_string)) AND (NOT (fv.value_boolean IS DISTINCT FROM f.value_boolean)))
         ->  Seq Scan on facts f  (cost=0.00..1027808.04 rows=40920604 width=98)
         ->  Hash  (cost=319709.05..319709.05 rows=367105 width=162)
               ->  Seq Scan on fact_values fv  (cost=0.00..319709.05 rows=367105 width=162)
(7 rows)
```

Which is going to take stupid amounts of time.

File: src/puppetlabs/puppetdb/scf/migrate.clj (hunk 3 above)
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: discussed in person but I bet someone has one that's 2mb. Separately, you may be able to do something like this: https://github.com/puppetlabs/puppetdb/blob/master/src/puppetlabs/puppetdb/scf/migrate.clj#L1140,L1147 here. That will use a fetchsize of 500

Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: on second thought, that's probably not useful for the batched update we need here

**Comment by wkalt** [comment]
generally seems sane. I'm going to work on a dataset for testing it now.

File: src/puppetlabs/puppetdb/scf/migrate.clj (hunk 2 above)
Comment (mullr) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: Updated this to handle nulls and non-nulls separately.

File: src/puppetlabs/puppetdb/scf/migrate.clj [hunk 4]
```diff
@@ -1281,13 +1283,121 @@
...
+               AND fv.value IS NOT DISTINCT FROM f.value
+        WHERE f.value IS NULL AND fv.value IS NULL")
+
+  ;; populate fact_values.value_hash
+  (log/info (trs "[4/6] Computing fact value hashes..."))
+  (jdbc/call-with-query-rows
+   ["select id, value::text from fact_values"]
+   (fn [rows]
+     (doseq [batch (partition-all 500 rows)]
+       (let [ids (map :id rows)
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/migrate.clj]:
```
(map :id batch)
```

File: src/puppetlabs/puppetdb/scf/migrate.clj [hunk 5]
```diff
@@ -1281,13 +1283,121 @@
...
+             hashes (map #(-> (:value %)
+                              json/parse
+                              hash/generic-identity-hash
+                              sutils/munge-hash-for-storage)
+                         rows)]
```
Comment (wkalt) [inline src/puppetlabs/puppetdb/scf/migrate.clj]: batch

**Comment by mullr** [comment]
Closing in favor of https://github.com/puppetlabs/puppetdb/pull/2346

**Pull request closed without merging**